src/crypto.rs

```diff
@@ -0,0 +1,217 @@
+    /// No encryption.
+    Null,
+    /// Encryption + authentication
+    Authenticated { shared_key: SharedSecretKey },
+    /// No message authentication. Only decrypt operation is allowed.
+    AnonymousDecrypt {
+        /// Our private key.
+        our_pk: PublicEncryptKey,
+        /// Our secret key.
+        our_sk: SecretEncryptKey,
+    },
+}
+
+impl Default for DecryptContext {
+    /// Default is "null" encryption.
+    fn default() -> Self {
+        Self::null()
+    }
+}
+
+impl DecryptContext {
+    /// Contructs "null" encryption context which actually does no encryption.
+    /// In this case data is simply serialized but not encrypted.
+    pub fn null() -> Self {
+        DecryptContext::Null
+    }
+
+    /// Construct crypto context that encrypts and authenticate messages.
+    pub fn authenticated(shared_key: SharedSecretKey) -> Self {
+        DecryptContext::Authenticated { shared_key }
+    }
+
+    /// Constructs crypto context that is only meant for unauthenticated decryption.
+    pub fn anonymous_decrypt(our_pk: PublicEncryptKey, our_sk: SecretEncryptKey) -> Self {
+        DecryptContext::AnonymousDecrypt { our_pk, our_sk }
+    }
+
+    /// Decrypt given buffer and deserialize into structure.
+    pub fn decrypt<T>(&self, msg: &[u8]) -> ::Res<T>
+    where
+        T: Serialize + DeserializeOwned,
+    {
+        Ok(match *self {
+            DecryptContext::Null => deserialise(msg)?,
+            DecryptContext::Authenticated { ref shared_key } => shared_key.decrypt(msg)?,
+            DecryptContext::AnonymousDecrypt {
+                ref our_pk,
+                ref our_sk,
+            } => our_sk.anonymously_decrypt(msg, our_pk)?,
+        })
+    }
+
+    /// The length of encrypted size variable. The returned value must match
+    /// `EncryptContext::encrypted_size_len()`, so that we could be able to decrypt it.
+    pub fn encrypted_size_len(&self) -> usize {
+        match *self {
+            // null encryption serializes 32 bit number into 4 byte array
+            DecryptContext::Null => 4,
+            // safe_crypto always serializes 32 bit number into 52 byte array
+            DecryptContext::Authenticated { .. } => 52,
+            DecryptContext::AnonymousDecrypt { .. } => 52,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use hamcrest2::prelude::*;
+    use safe_crypto::gen_encrypt_keypair;
+    use std::u32::MAX as MAX_U32;
+    use DEFAULT_MAX_PAYLOAD_SIZE;
+
+    mod encrypt_context {
+        use super::*;
+
+        #[test]
+        fn encrypt_always_returns_52_byte_array_for_4_byte_input_with_anonymous_encryption() {
+            let (pk, _sk) = gen_encrypt_keypair();
+            let enc_ctx = EncryptContext::anonymous_encrypt(pk);
+
+            for size in &[0u32, 25000, DEFAULT_MAX_PAYLOAD_SIZE as u32, MAX_U32] {
+                let encrypted = unwrap!(enc_ctx.encrypt(&size));
+                assert_that!(&encrypted, len(52));
```

Comment:
Use global constant from production as in review comment above. Same with other cases below.